src/Tasks/SystemState.cs

```diff
@@ -544,12 +549,73 @@ out fileState.frameworkName
         }
 
         /// <summary>
-        /// Cached implementation of GetDirectories.
+        /// Reads in cached data from stateFiles to build an initial cache. Avoids logging warnings or errors.
         /// </summary>
-        /// <param name="path"></param>
-        /// <param name="pattern"></param>
-        /// <returns></returns>
-        private string[] GetDirectories(string path, string pattern)
+        internal static SystemState DeserializePrecomputedCaches(ITaskItem[] stateFiles, TaskLoggingHelper log, Type requiredReturnType, GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo, Func<string, bool> fileExists)
+        {
+            SystemState retVal = new SystemState();
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+            fileExists ??= FileSystems.Default.FileExists;
+
+            foreach (ITaskItem stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sysState = DeserializeCache<SystemState>(stateFile.ToString(), log, false);
```

Comment:
I hadn't been thinking about that. The way I've been using them is passing in full paths, and users could get those from property + relative path if necessary. It's certainly limiting to make this assumption, but since it's new functionality, no one will be broken by it being restricted in that way, and it would certainly be faster to just know that it's a full path to the cache. I could go either way. It sounds like you'd prefer it account for globs and paths from other projects? We might need to talk about that if so, since I don't think I know all of how to do that.